// sdk/models/User.js
/**
 * «mobileSimbirsoft Platform».
 * Интеллектуальная собственность ООО «Мобайл СимбирСофт».
 * Copyright © 2017 by «Мобайл СимбирСофт».
 */

'use strict';

const md5 = require('md5');
const co = require('co');

const Adapter = require('./DBAdapter');
const DBAdapter = Adapter.get();

const symbols = {};
const sym = name => (symbols[name] || (symbols[name] = Symbol(name)));
const [ required ] = [true];

class User {
  constructor(data = {}, application) {
    this[sym('model')] = data;
    this[sym('application')] = application;
  }

  static get schema() {
    return {
      username: { type: 'string', required },
      password: { type: 'string' },
      role: { type: 'id', default: null, referer: 'Role', model: require('./Role') },
      type: { type: 'string', default: 'oauth' },
    };
  }

  static get scemaOptions() {
    return {
      timestamps: true,
      index: [
        [['username', 'type'], { unique: true }],
      ]
    };
  }

  static findAll(params, options = {}, application) {
    return co(function *() {
      let tableName = application ? `User_${application}` : 'User';
      let users = yield DBAdapter.init(tableName, this.schema, this.scemaOptions, application).findAll(params, options);
      return (users || []).map(u => new this(u, application));
    }.bind(this));
  }

  static find(params, options = {}, application) {
    return co(function *() {
      let tableName = application ? `User_${application}` : 'User';
      let user = yield DBAdapter.init(tableName, this.schema, this.scemaOptions, application).find(params, options);
      return user ? new this(user, application) : null;
    }.bind(this));
  }

  static cryptPassword(password) {
    return md5(password);
  }

  static access(req, rule) {
    if (rule) return ['user', rule];

    switch (req.method.toUpperCase()) {
      case 'GET': return req.params && req.params.user_id ? ['user', 'view'] : ['user', 'list'];
      case 'POST': return ['user', 'create'];
      case 'PUT': return ['user', 'update'];
      case 'DELETE': return ['user', 'delete'];
    }
  }

  static validatePermission(name, access, req) {
    switch (name) {
      case 'all':
        return ['list', 'create'].indexOf(access) > -1;

      case 'own':
        if (['list', 'view', 'update', 'delete'].indexOf(access) > -1) {
          return {
            filter: {id: req.user.id}
          }
        }
        return false;

      default: return false;
    }
  }

  validatePermission(name, access, req) {
    switch (name) {
      case 'all':
        return ['list', 'view', 'create', 'update', 'delete'].indexOf(access) > -1;

      case 'own':
        if (this.id && this.id.toString() == req.user.id.toString()) {
          if (['list', 'view', 'update', 'delete'].indexOf(access) > -1) {
            return {
              filter: {id: req.user.id},
              select: ['username', 'role', 'type', 'createdAt', 'updatedAt']
            }
          }
          return false;
        } else {
          return false;
        }

      default: return false;
    }
  }

  get id() { return this.self.id ? this.self.id.toString() : null; }
  get username() { return this.self.username; }
  get password() { return this.self.password; }
  get type() { return this.self.type; }
  get role() { return this.self.role; }
  get application() { return this[sym('application')]; }

  get self() { return this[sym('model')]; }

  set username(value) { this.self.username = value; }
  set password(value) {
    if (this.self.password != value) {
      this.self.password = this.constructor.cryptPassword(value);
    }
  }
  set type(value) { this.self.type = value; }
  set role(value) { this.self.role = value; }
  set id(value) {};

  save() {
    return co(function *() {
      let user;
      let schema = this.constructor.schema;
      let options = this.constructor.schemaOptions;
      let tableName = this.application ? `User_${this.application}` : 'User';
      if (this.self.role === '') this.self.role = undefined;

      if (this.id) {
        user = yield DBAdapter.init(tableName, schema, options).update(this.self);
      } else {
        user = yield DBAdapter.init(tableName, schema, options).insert(this.self);
      }

      this.self.id = user.id;

      return this;
    }.bind(this));
  }

  update(newData) {
    return co(function *() {
      Object.keys(newData).map(f => this[f] = newData[f]);
      return this.save();
    }.bind(this));
  }

  remove() {
    return co(function *() {
      let schema = this.constructor.schema;
      let options = this.constructor.schemaOptions;
      let tableName = this.application ? `User_${this.application}` : 'User';
      yield DBAdapter.init(tableName, schema, options).remove({ id: this.id });
      Object.keys(this.self).map(k => delete this.self[k]);
      return this;
    }.bind(this));
  }
}

module.exports = User;
