// sdk/models/auth/OAuth2.js
/**
 * «mobileSimbirsoft Platform».
 * Интеллектуальная собственность ООО «Мобайл СимбирСофт».
 * Copyright © 2017 by «Мобайл СимбирСофт».
 */

'use strict';

var http = require('http');
var oauthServer = require('oauth2-server');
const co = require('co');

const AFError = require('../Error');
const User = require('../User');
const Application = require('../Application');
const Entity = require('../Entity');
const Adapter = require('../DBAdapter');
const DBAdapter = Adapter.get();

const symbols = {};
const sym = name => (symbols[name] || (symbols[name] = Symbol(name)));
const [required, array] = [true, true];

class OAuth {
  constructor(options) {
    this.options = options;

    this.server = oauthServer(Object.assign({
      model: this,
      continueAfterResponse: true,
      grants: options.grants || ['password', 'refresh_token'],
      sms_code: options.sms_code
    }, options.tokenOptions ? options.tokenOptions : this.defaultOptions, this.options));
  }

  static get AccessToken() { return AccessToken; }
  static get RefreshToken() { return RefreshToken; }

  get defaultOptions() {
    return {
      accessTokenLifetime: 60 * 60 * 24,
      refreshTokenLifetime: 60 * 60 * 24 * 7,
    };
  }

  setReq(req) { this.req = req; return this; }

  auth() {
    return new Promise((resolve, reject) => {
      if (this.req.request && this.req.request.fields) {
        this.req.body = Object.assign({}, this.req.request.fields);
      }

      this.res = {
        redirect: (uri) => {
          resolve(uri);
        }
      };

      this.server.authCodeGrant(this.check)(this.req, this.res, (err) => {
        if (err) { reject(err); }
      });
    });
  }

  login() {
    return new Promise((resolve, reject) => {
      if (this.req.request && this.req.request.fields) {
        this.req.body = Object.assign({}, this.req.request.fields);
      }

      this.req.headers['content-type'] = 'application/x-www-form-urlencoded';
      this.req.body.grant_type = 'authorization_code';

      this.res = {
        set: () => { },
        jsonp: (token) => {
          token.user_id = this.req.user.id;

          if (this.req.user.role && !this.req.user.role.permissions) {
            return reject(new AFError(403, 'Permission denied'));
          }

          resolve(token);
        }
      };

      this.server.grant()(this.req, this.res, (err) => {
        if (err) { reject(err); }
      });
    });
  }


  logout() {
    return co(function* () {
      if (this.req.oauth && this.req.oauth.bearerToken) {
        let accessTokensTable = DBAdapter.init('AccessToken', AccessToken.schema);
        yield accessTokensTable.remove({
          clientId: this.req.oauth.bearerToken.clientId,
          accessToken: this.req.oauth.bearerToken.accessToken
        });

        let refreshTokensTable = DBAdapter.init('RefreshToken', RefreshToken.schema);
        yield refreshTokensTable.remove({
          clientId: this.req.oauth.bearerToken.clientId,
          refreshToken: this.req.oauth.bearerToken.refreshToken
        });
      } else {
        throw new AFError(401, 'Unauthorized');
      }
    }.bind(this));
  }

  refresh() {
    return new Promise((resolve, reject) => {
      this.req.headers['content-type'] = 'application/x-www-form-urlencoded';
      this.req.body = this.req.body || {};
      if (this.req.request && this.req.request.fields) {
        this.req.body = this.req.request.fields;
      }
      this.req.body.grant_type = 'refresh_token';

      let req = this.req;
      this.res = {
        set: () => { },
        jsonp: (token) => {
          token.user_id = req.user.id;
          resolve(token);
        }
      };

      this.server.grant()(this.req, this.res, (err) => {
        if (err) { reject(err); }
      });
    });
  }

  getAccessToken(accessToken, callback) {
    co(function* () {
      let token = yield DBAdapter.init('AccessToken', AccessToken.schema).find({ accessToken });
      if (!token) callback(new AFError(401, 'Access token not found'));
      callback(null, token);
    }.bind(this));
  }

  saveAccessToken(accessToken, clientId, expires, user, callback) {
    co(function* () {
      let data = {
        accessToken: accessToken,
        application: this.req.application ? this.req.application.id : null,
        clientId: clientId,
        userId: user.id,
        expires: expires,
      };

      if (this.req.application) { data.application = this.req.application.id; }

      let token = yield DBAdapter.init('AccessToken', AccessToken.schema).insert(data);
      if (!token) callback(new AFError(500, 'Access token not saved'));
      callback(null, token);
    }.bind(this));
  }

  getClient(clientId, clientSecret, callback) {
    if (clientSecret) {
      if (this.req.application && this.req.application.secret !== clientSecret) {
        return callback(new AFError(401, 'Invalid clientSecret'));
      }

      callback(null, { clientId: clientId, clientSecret: clientSecret });
    } else {
      callback(null, { clientId: clientId, redirectUri: this.options.redirectUri });
    }
  }

  check(req, callback) {
    co(function* () {
      let userPromise = User.find({
        username: req.body.username,
      }, { populate: 'role' }, this.req.application ? this.req.application.id : null);

      let user = yield userPromise;

      if (user) {
        return callback(null, true, user);
      }
      callback(new AFError(401, 'User not found'));
    }.bind(this));
  }

  grantTypeAllowed(clientId, grantType, callback) {
    if (['authorization_code', 'refresh_token'].indexOf(grantType) >= 0) {
      return callback(null, true);
    }

    callback(null, false);
  }

  getUser(username, password, callback) {
    co(function* () {
      let user = yield User.find({
        username: username,
        password: User.cryptPassword(password),
      }, { populate: 'role' }, this.req.application ? this.req.application.id : null);

      if (user) {
        return callback(null, user);
      }

      callback(new AFError(401, 'User not found'));
    }.bind(this));
  }

  getAuthCode(authCode, callback) {
    return co(function* () {
      let code = yield DBAdapter.init('AuthorizationCode', AuthorizationCode.schema).find({ authCode });
      if (!callback) return code;
      if (!code) callback(new AFError(401, 'Authorization code not found'));
      callback(null, code);
    }.bind(this));
  }

  doAuthCodeTrying(authCode, limit) {
      co(function* () {
          let code = yield DBAdapter.init('AuthorizationCode', AuthorizationCode.schema).find({ authCode });
          if(code.trying >= limit) {
              let res = yield DBAdapter.init('AuthorizationCode', AuthorizationCode.schema).remove({ authCode });
          } else {
              let res = yield DBAdapter.init('AuthorizationCode', AuthorizationCode.schema)
                  .findOneAndUpdate(
                      {authCode},
                      {trying: (code.trying || 0) + 1},
                      {}
                  );
          }
      }.bind(this));
  }

  getAllAuthCodes(param) {
    return co(function* () {
      let codes = yield DBAdapter.init('AuthorizationCode', AuthorizationCode.schema).findAll(param);
      return codes;
    }.bind(this));
  }

  saveAuthCode(authCode, clientId, expires, user, callback) {
    co(function* () {
      var data = {
        authCode: authCode,
        clientId: clientId,
        userId: user.id,
        expires: expires,
        smsCode: this.options.sms_code,
        trying: 0
      };

      if (this.req.application) { data.application = this.req.application.id; }

      let code = yield DBAdapter.init('AuthorizationCode', AuthorizationCode.schema).insert(data);
      if (!code) callback(new AFError(500, 'Auth code not saved'));
      callback(null, code);
    }.bind(this));
  }

  saveRefreshToken(refreshToken, clientId, expires, user, callback) {
    co(function* () {
      var data = {
        refreshToken: refreshToken,
        clientId: clientId,
        userId: user.id,
        expires: expires,
      };
      if (this.req.application) { data.application = this.req.application.id; }

      let token = yield DBAdapter.init('RefreshToken', RefreshToken.schema).insert(data);
      if (!token) callback(new AFError(500, 'Refresh token not saved'));
      callback(null, token);
    }.bind(this));
  }

  getRefreshToken(refreshToken, callback) {
    co(function* () {
      let token = yield DBAdapter.init('RefreshToken', RefreshToken.schema).find({ refreshToken });
      if (!token) callback(new AFError(401, 'Refresh token not found'));
      callback(null, token);
    }.bind(this));
  }
}

class AuthorizationCode {
  static get schema() {
    return {
      authCode: { type: 'string' },
      smsCode: { type: 'string' },
      application: { type: 'id', referer: 'Application', model: require('../Application') },
      clientId: { type: 'string' },
      userId: { type: 'id', referer: 'User', model: require('../User') },
      expires: { type: 'date', expires: 3600 },
      trying: { type: 'integer' }
    };
  }
}

class AccessToken {
  static get schema() {
    return {
      accessToken: { type: 'string' },
      application: { type: 'id', referer: 'Application', model: require('../Application') },
      clientId: { type: 'string' },
      userId: { type: 'id', referer: 'User', model: require('../User') },
      expires: { type: 'date', expires: 3600 }
    };
  }
}

class RefreshToken {
  static get schema() {
    return {
      refreshToken: { type: 'string' },
      application: { type: 'id', referer: 'Application', model: require('../Application') },
      clientId: { type: 'string' },
      userId: { type: 'id', referer: 'User', model: require('../User') },
      expires: { type: 'date', expires: 3600 }
    };
  }
}

module.exports = OAuth;
